### ISSUE TITLE
Cannot generate colon after `else` in Python

### ISSUE THREAD

**Issue by mindol** [uiuc-focal-lab/syncode#140]
## Problem
Hello, I think  SynCode cannot generate a **colon** after `else` properly in Python.
Currently, I'm using `microsoft/phi-2` model and tokenizer.

For example, after the following code, "return" is generated while ":"(colon) is expected.
```python
def is_palindrome(n):
  if str(n) == str(n)[::-1]:
    return True
  else
```
It becomes:
```python
def is_palindrome(n):
  if str(n) == str(n)[::-1]:
    return True
  elsereturn
```

## Example
Here is the minimal reproduce example:
(It is quite long, but it just fetches the scores of the one-token inference)

**Dependencies**:
syncode v0.1.4
torch==2.5.1
transformers==4.44.0

**Example Code**:
```python
from syncode.grammar_decoder import SyncodeLogitsProcessor
from syncode.parsers.grammars import Grammar
import torch
from transformers import AutoModelForCausalLM
from transformers import AutoTokenizer

# Step 1. Load model and tokenizer
torch.set_default_device("cpu")

model_name = "microsoft/phi-2"
model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Step 2. Set prompt
prompt = """def is_palindrome(n):
  if str(n) == str(n)[::-1]:
    return True
  else"""

# Step 3. Initialize SyncodeLogitsProcessor
syncode_processor = SyncodeLogitsProcessor(grammar=Grammar("python"),
                                           tokenizer=tokenizer,
                                           parse_output_only=False)
syncode_processor.reset(prompt)

# Step 4. Generate scores with SyncodeLogitsProcessor
inputs = tokenizer(prompt, return_tensors="pt").input_ids
outputs = model.generate(inputs,
                         attention_mask=torch.ones_like(inputs),
                         do_sample=True,
                         logits_processor=[syncode_processor],
                         return_dict_in_generate=True,
                         output_scores=True,
                         max_new_tokens=1)

# Step 5. Print scores
scores = outputs.scores
generated_id = outputs.sequences[0, -1]
generated_str = tokenizer.decode(generated_id, skip_special_tokens=True)
print(f"Generated token id: {generated_id} ({generated_str})")
print(f"Score of this token: {scores[0][-1][generated_id]}")

colon_id = tokenizer.encode(':')[0]
colon_str = tokenizer.decode(colon_id)
print(f"Colon token id: {colon_id} ({colon_str})")
print(f"Score of colon token: {scores[0][-1][colon_id]}")
```
**Output**:
```bash
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:02<00:00,  1.16s/it]
Creating DFA mask store for CodeGenTokenizerFast and python, may take more than 10 minutes. Caching at /home/mindol/.......
Ignore whitespace tokens is True
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 402/402 [12:18<00:00,  1.84s/it]
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Generated token id: 7783 (return)
Score of this token: 13.40625
Colon token id: 25 (:)
Score of colon token: -inf
```
**Explanation**:
Score of colon token is `-inf`, which should not be.

**Comment by shubhamugare** [comment]
Hi @mindol,

Thanks for filing the issue. It should be fixed in the latest PR #141. I have added a few tests to check this as well.

**Comment by mindol** [comment]
@shubhamugare Thank you very much! Now it works well on the example code.
```bash
Generated token id: 25 (:)  # colon generated
Score of this token: 28.345806121826172
Colon token id: 25 (:)
Score of colon token: 28.345806121826172
```
